File: src/doctor.ts
// doctor is the safety net for "files are truth": it compares the index to the
// files, repairs what drifted, and reports what only a human can fix.
import type { Database } from "bun:sqlite";
import { renameSync, rmSync } from "node:fs";
import { openDb, dbPath } from "./db";
import { reindex, readNote, scanVault } from "./indexer";
import { linkTarget } from "./note";
import type { Embedder } from "./embed";

/** An unresolved link: `slug` is the target as written — bare stem or path. */
export type LinkProblem = { from: string; slug: string };

/**
 * A bare-stem link several notes answer to. `candidates` are link *targets*,
 * not filenames (`customers/acme`, no `.md`), so writing one of them into the
 * note as `[[customers/acme]]` is the whole fix. A candidate at the vault root
 * has no qualified form and reads as the ambiguous stem itself — that note has
 * to move into a namespace instead.
 */
export type AmbiguousLink = LinkProblem & { candidates: string[] };

export type DoctorReport = {
  /** files whose indexed hash is wrong or missing (reindexed when repairing) */
  stale: string[];
  /** indexed rows whose file is gone (purged when repairing) */
  missing: string[];
  /** `[[target]]` matching no note at all — a typo, or a note that is gone */
  brokenLinks: LinkProblem[];
  /** `[[stem]]` matching several notes — none of them wins; the paths that did */
  ambiguousLinks: AmbiguousLink[];
  /** notes with no *resolved* link in and no link out at all */
  orphans: string[];
  /** frontmatter the parser could not use — indexed anyway, whole file as body */
  malformed: string[];
  /** every note was re-embedded: the model or dims changed, or this was a rebuild */
  reembedded: boolean;
};

export type DoctorOptions = {
  /** reindex stale notes and purge deleted ones (default true) */
  repair?: boolean;
  /** index into a temp DB and rename it over index.db (instead of repairing) */
  rebuild?: boolean;
};

export async function doctor(
  root: string,
  embedder: Embedder,
  { repair = true, rebuild = false }: DoctorOptions = {},
): Promise<DoctorReport> {
  const path = dbPath(root);
  // Drift is measured before any repair, so the report says what was wrong.
  const drift = await withDb(path, (db) => diskDrift(db, root));
  let reembedded = rebuild; // a rebuild embeds every note from scratch
  if (rebuild) await rebuildIndex(root, embedder);
  else if (repair) {
    reembedded = (await withDb(path, (db) => reindex(db, root, embedder))).reembedded;
  }
  return { ...drift, ...(await withDb(path, graphReport)), reembedded };
}

async function withDb<T>(path: string, fn: (db: Database) => T | Promise<T>): Promise<T> {
  const db = openDb(path);
  try {
    return await fn(db);
  } finally {
    db.close();
  }
}

/**
 * What the files say that the index does not.
 * ponytail: re-reads and re-hashes every note, and a repair then reads them
 * again. Fine for a hand-written vault; gate on mtime if one gets huge.
 */
async function diskDrift(
  db: Database,
  root: string,
): Promise<{ stale: string[]; missing: string[] }> {
  const indexed = new Map(
    (db.query(`select path, hash from notes`).all() as { path: string; hash: string }[]).map(
      (r) => [r.path, r.hash],
    ),
  );
  const stale: string[] = [];
  for (const rel of scanVault(root)) {
    const note = await readNote(root, rel);
    if (note === null) continue; // deleted while we looked: it counts as missing
    if (indexed.get(rel) !== note.hash) stale.push(rel);
    indexed.delete(rel);
  }
  return { stale, missing: [...indexed.keys()].sort() };
}

function graphReport(db: Database): Omit<DoctorReport, "stale" | "missing" | "reembedded"> {
  const unresolved = db
    .query(
      `select n.path as "from", e.to_slug as slug
       from edges e join notes n on n.id = e.from_id
       where e.to_id is null
       order by n.path, e.to_slug`,
    )
    .all() as LinkProblem[];

  // Stems several notes share. That is legitimate — namespaces are what they
  // are for — so it is not itself reported; it only matters as the candidate
  // list of a *bare* link that lands on one.
  const shared = new Map<string, string[]>();
  const dupes = db
    .query(
      `select slug, path from notes
       where slug in (select slug from notes group by slug having count(*) > 1)
       order by slug, path`,
    )
    .all() as { slug: string; path: string }[];
  for (const { slug, path } of dupes) {
    // stored as the link that would fix the note, not as the filename
    const target = linkTarget(path);
    const targets = shared.get(slug);
    if (targets) targets.push(target);
    else shared.set(slug, [target]);
  }

  const brokenLinks: LinkProblem[] = [];
  const ambiguousLinks: AmbiguousLink[] = [];
  for (const link of unresolved) {
    // A path-qualified target matches one note or none: it is never ambiguous.
    const candidates = link.slug.includes("/") ? undefined : shared.get(link.slug);
    if (candidates) ambiguousLinks.push({ ...link, candidates: [...candidates] });
    else brokenLinks.push(link);
  }

  const paths = (sql: string): string[] =>
    (db.query(sql).all() as { path: string }[]).map((r) => r.path);

  return {
    brokenLinks,
    ambiguousLinks,
    orphans: paths(`select path from notes n
       where not exists (select 1 from edges e where e.from_id = n.id)
         and not exists (select 1 from edges e where e.to_id = n.id)
       order by path`),
    malformed: paths(`select path from notes where malformed = 1 order by path`),
  };
}

/**
 * Rebuild from the files into a temp DB, then rename it over index.db — a
 * half-finished rebuild can never become the live index, and a failed one
 * leaves nothing behind.
 * ponytail: two races remain, both narrowed rather than closed. A watcher
 * holding the old file keeps writing to the replaced inode: those writes are
 * lost, not corrupting, and the next doctor run picks them back up off disk.
 * And a watcher that recreates the old file's -wal between our unlink and the
 * rename leaves a stale journal beside the fresh index, which SQLite would
 * then try to apply. A lock file is the upgrade path for both.
 */
async function rebuildIndex(root: string, embedder: Embedder): Promise<void> {
  const target = dbPath(root);
  const tmp = `${target}.rebuild-${Bun.randomUUIDv7()}`;
  try {
    await withDb(tmp, async (db) => {
      await reindex(db, root, embedder);
      db.run("pragma wal_checkpoint(truncate)"); // fold the WAL in before the rename
    });
  } catch (e) {
    rmDb(tmp); // a dead temp index must not accumulate in .vault/
    throw e;
  }
  // The replaced file's journal describes a database that is about to stop
  // existing; drop it first so it is never seen next to the new one.
  rmDb(target, { journalsOnly: true });
  renameSync(tmp, target);
}

function rmDb(path: string, { journalsOnly = false } = {}): void {
  for (const suffix of ["-wal", "-shm"]) rmSync(`${path}${suffix}`, { force: true });
  if (!journalsOnly) rmSync(path, { force: true });
}

File: src/wilcus_vault/doctor.py
"""The safety net for files-are-truth: compare the index to the files, repair
what drifted, and report what only a human can fix."""

import sqlite3
import uuid
from collections.abc import Callable
from dataclasses import dataclass, field
from pathlib import Path

from .db import db_path, open_db
from .discard_log import append_nofollow, discard_log, ensure_gitignore, read_nofollow
from .discards import count_discards
from .embed import Embedder
from .indexer import read_note, reindex, scan_vault
from .note import link_target


@dataclass(frozen=True)
class LinkProblem:
    """An unresolved link. `slug` is the target as written: a bare stem or a path."""

    from_path: str
    slug: str


@dataclass(frozen=True)
class AmbiguousLink(LinkProblem):
    """A bare-stem link several notes answer to. `candidates` are link targets
    (`customers/acme`), so writing one into the note is the whole fix."""

    candidates: list[str] = field(default_factory=list)


@dataclass
class DoctorReport:
    stale: list[str]  # files whose indexed hash is wrong or missing
    missing: list[str]  # indexed rows whose file is gone
    broken_links: list[LinkProblem]  # `[[target]]` matching no note at all
    ambiguous_links: list[AmbiguousLink]  # `[[stem]]` matching several notes
    orphans: list[str]  # no resolved link in and no link out
    malformed: list[str]  # frontmatter the parser could not use
    reembedded: bool  # every note was re-embedded: model or dims changed, or a rebuild
    migrated_discard_log: bool  # a log left in `.vault/` was moved beside the notes
    discards: dict[str, int]  # discard log: total entries, and those from the last 7 days


@dataclass(frozen=True)
class DoctorOptions:
    repair: bool = True  # reindex stale notes and purge deleted ones
    rebuild: bool = False  # index into a temp DB and rename it over index.db


async def doctor(
    root: str | Path, embedder: Embedder, options: DoctorOptions | None = None
) -> DoctorReport:
    opts = options or DoctorOptions()
    root = Path(root)
    path = db_path(root)
    # First, before anything touches `.vault/`: history still in there is one
    # `rm -rf .vault` from gone. Only on a repairing run; a report moves nothing.
    migrated = (opts.repair or opts.rebuild) and _migrate_discard_log(root)
    # Drift is measured before any repair, so the report says what was wrong.
    stale, missing = _with_db(path, lambda db: _disk_drift(db, root))
    reembedded = opts.rebuild
    if opts.rebuild:
        await _rebuild_index(root, embedder)
    elif opts.repair:
        db = open_db(path)
        try:
            reembedded = (await reindex(db, root, embedder)).reembedded
        finally:
            db.close()
    broken, ambiguous, orphans, malformed = _with_db(path, _graph_report)
    return DoctorReport(
        stale=stale,
        missing=missing,
        broken_links=broken,
        ambiguous_links=ambiguous,
        orphans=orphans,
        malformed=malformed,
        reembedded=reembedded,
        migrated_discard_log=migrated,
        discards=count_discards(root),
    )


def _migrate_discard_log(root: Path) -> bool:
    """Move a discard log out of the disposable `.vault/` directory, once. Appended
    rather than replaced: both files may hold lines and neither is disposable.
    ponytail: append-then-remove is not atomic; a crash between duplicates lines
    on the next run, never loses them."""
    old = root / ".vault" / "discarded.log"
    if not old.exists():
        return False
    lines = read_nofollow(old)
    if lines:
        log = discard_log(root)
        if not log.exists():
            ensure_gitignore(root)
        append_nofollow(log, lines if lines.endswith("\n") else lines + "\n")
    old.unlink()
    return True


def _with_db[T](path: Path, fn: Callable[[sqlite3.Connection], T]) -> T:
    db = open_db(path)
    try:
        return fn(db)
    finally:
        db.close()


def _disk_drift(db: sqlite3.Connection, root: Path) -> tuple[list[str], list[str]]:
    """(stale, missing): what the files say that the index does not.
    ponytail: re-reads and re-hashes every note; gate on mtime if a vault gets huge."""
    indexed = {r["path"]: r["hash"] for r in db.execute("select path, hash from notes")}
    stale = []
    for rel in scan_vault(root):
        note = read_note(root, rel)
        if note is None:
            continue  # deleted while we looked: it counts as missing
        if indexed.get(rel) != note.hash:
            stale.append(rel)
        indexed.pop(rel, None)
    return stale, sorted(indexed)


def _graph_report(
    db: sqlite3.Connection,
) -> tuple[list[LinkProblem], list[AmbiguousLink], list[str], list[str]]:
    unresolved = db.execute(
        """select n.path as from_path, e.to_slug as slug
           from edges e join notes n on n.id = e.from_id
           where e.to_id is null order by n.path, e.to_slug"""
    ).fetchall()
    # Stems several notes share: legitimate on its own, only reported as the
    # candidate list of a bare link that lands on one.
    shared: dict[str, list[str]] = {}
    for row in db.execute(
        """select slug, path from notes
           where slug in (select slug from notes group by slug having count(*) > 1)
           order by slug, path"""
    ):
        shared.setdefault(row["slug"], []).append(link_target(row["path"]))

    broken: list[LinkProblem] = []
    ambiguous: list[AmbiguousLink] = []
    for row in unresolved:
        # A path-qualified target matches one note or none: never ambiguous.
        candidates = None if "/" in row["slug"] else shared.get(row["slug"])
        if candidates:
            ambiguous.append(AmbiguousLink(row["from_path"], row["slug"], list(candidates)))
        else:
            broken.append(LinkProblem(row["from_path"], row["slug"]))

    orphans = [
        r["path"]
        for r in db.execute(
            """select path from notes n
               where not exists (select 1 from edges e where e.from_id = n.id)
                 and not exists (select 1 from edges e where e.to_id = n.id)
               order by path"""
        )
    ]
    malformed = [
        r["path"] for r in db.execute("select path from notes where malformed = 1 order by path")
    ]
    return broken, ambiguous, orphans, malformed


async def _rebuild_index(root: Path, embedder: Embedder) -> None:
    """Rebuild from the files into a temp DB, then rename it over index.db: a
    half-finished rebuild can never become the live index.
    ponytail: a watcher holding the old file keeps writing to the replaced inode;
    those writes are lost, not corrupting, and the next doctor run recovers them."""
    target = db_path(root)
    tmp = target.with_name(f"{target.name}.rebuild-{uuid.uuid4().hex}")
    db = open_db(tmp)
    try:
        await reindex(db, root, embedder)
        db.execute("pragma wal_checkpoint(truncate)")  # fold the WAL in before the rename
    except BaseException:
        db.close()
        _rm_db(tmp)
        raise
    db.close()
    _rm_db(target, journals_only=True)  # the old journal describes a database about to vanish
    tmp.replace(target)


def _rm_db(path: Path, journals_only: bool = False) -> None:
    for suffix in ("-wal", "-shm"):
        path.with_name(path.name + suffix).unlink(missing_ok=True)
    if not journals_only:
        path.unlink(missing_ok=True)
